File: Python/conway.py
from time import sleep


class Conway:
    def __init__(self, height, width):
        self.height = height
        self.width = width
        self.grid = self.new_grid()

    def new_grid(self):
        return [[False for _ in range(0, self.width)]
                for _ in range(0, self.height)]

    def query(self, x, y):
        return self.grid[x % self.height][y % self.width]

    def assign(self, x, y, state):
        self.grid[x % self.height][y % self.width] = state

    def count_neighbors(self, x, y):
        n  = self.query(x + 1, y    )
        ne = self.query(x + 1, y + 1)
        e  = self.query(x,     y + 1)
        se = self.query(x - 1, y + 1)
        s  = self.query(x - 1, y    )
        sw = self.query(x - 1, y - 1)
        w  = self.query(x,     y - 1)
        nw = self.query(x + 1, y - 1)
        neighbor_states = [n, ne, e, se, s, sw, w, nw]
        return len(list(filter(lambda state: state, neighbor_states)))

    def draw(self, value):
        return '#' if value else '.'

    def print_grid(self):
        for row in self.grid:
            for c in row:
                print(self.draw(c), end='')
            print()

    def game_logic(self, state, neighbors):
        match state:
            case True:
                if neighbors < 2:
                    return False
                if neighbors > 3:
                    return False
                return state
            case False:
                if neighbors == 3:
                    return True
                return state

    def step_cell(self, x, y):
        state = self.query(x, y)
        neighbors = self.count_neighbors(x, y)
        next_state = self.game_logic(state, neighbors)
        return next_state

    def simulate(self):
        new_g = self.new_grid()

        for x in range(0, self.height):
            for y in range(0, self.width):
                new_g[x][y] = self.step_cell(x, y)

        self.grid = new_g

    def clear_screen(self):
        print('\x1b[2J')


if __name__ == '__main__':
    con = Conway(10, 20)

    con.assign(0, 3, True)
    con.assign(1, 4, True)
    con.assign(2, 2, True)
    con.assign(2, 3, True)
    con.assign(2, 4, True)

    con.print_grid()

    while True:
        con.clear_screen()
        con.simulate()
        con.print_grid()
        sleep(0.1)

# python3 conway.py

File: Go/conway.go
package main

import (
	"fmt"
	"time"
)

const HEIGHT = 10
const WIDTH = 20

type Grid = [HEIGHT][WIDTH]int

func Mod(x int, y int) int {
	if x > 0 {
		return x % y
	}
	if x < 0 {
		return y + x%y
	}
	return 0
}

func Query(grid Grid, x int, y int) int {
	return grid[Mod(x, HEIGHT)][Mod(y, WIDTH)]
}

func Assign(grid *Grid, x int, y int, state int) {
	grid[Mod(x, HEIGHT)][Mod(y, WIDTH)] = state
}

func CountNeighbors(grid Grid, x int, y int) int {
	var n  = Query(grid, x + 1, y    )
	var ne = Query(grid, x + 1, y + 1)
	var e  = Query(grid, x,     y + 1)
	var se = Query(grid, x - 1, y + 1)
	var s  = Query(grid, x - 1, y    )
	var sw = Query(grid, x - 1, y - 1)
	var w  = Query(grid, x,     y - 1)
	var nw = Query(grid, x + 1, y - 1)
	return n + ne + e + se + s + sw + w + nw
}

func Draw(value int) byte {
	if value == 1 {
		return '#'
	} else {
		return '.'
	}
}

func Print(grid Grid) {
	for i := range grid {
		for j := range grid[i] {
			fmt.Printf("%c", Draw(grid[i][j]))
		}
		fmt.Println()
	}
}

func GameLogic(state int, neighbors int) int {
	if state == 1 {
		if neighbors < 2 {
			return 0
		}
		if neighbors > 3 {
			return 0
		}
		return state
	} else {
		if neighbors == 3 {
			return 1
		}
		return state
	}
}

func StepCell(grid Grid, x int, y int) int {
	var state = Query(grid, x, y)
	var neighbors = CountNeighbors(grid, x, y)
	var nextState = GameLogic(state, neighbors)
	return nextState
}

func Simulate(grid Grid) Grid {
	var newGrid Grid
	for x := range grid {
		for y := range grid[x] {
			newGrid[x][y] = StepCell(grid, x, y)
		}
	}
	return newGrid
}

func ClearScreen() {
	fmt.Println("\x1B[2J")
}

func main() {
	var grid Grid

	Assign(&grid, 0, 3, 1)
	Assign(&grid, 1, 4, 1)
	Assign(&grid, 2, 2, 1)
	Assign(&grid, 2, 3, 1)
	Assign(&grid, 2, 4, 1)

	Print(grid)

	for {
		ClearScreen()
		grid = Simulate(grid)
		Print(grid)
		time.Sleep(100 * time.Millisecond)
	}
}

// go run conway.go
